### ISSUE TITLE
Points

### ISSUE THREAD

**Issue by cristeigabriel** [NukoOoOoOoO/winter-linux#3]
https://github.com/NukoOoOoOoO/winter-linux/blob/1d6a8806d732250ed925d43cb4cfd5f4eb424c7b/src/sdk/interfaces/localize.h#L7

You can use FindUtf8 directly from localizer. [HelvetaCS does this](https://github.com/cristeigabriel/HelvetaCS/) so then you won't have to do UCS->UTF-8. If you need help - I can implement it, although I don't really use Linux for gaming.

https://github.com/NukoOoOoOoO/winter-linux/blob/e1c915ceb5d438a8bf55258a73b95aa6c0c679f3/src/utils/memory.cpp#L15

You can just hold instances of all modules as some class' objects and have them inherit stuff like pattern scanner as some sort of traits. You could probably do some concept hackery with this if you want to do it as I said.

https://github.com/NukoOoOoOoO/winter-linux/blob/e1c915ceb5d438a8bf55258a73b95aa6c0c679f3/src/utils/memory.cpp#L42

This is unneeded unless you plan on networking strings for signature scanning. You can use [STB](https://github.com/cristeigabriel/STB) to compile strings to byte arrays at compile-time. It'd require that you make a template and pass size there to use an array - or - the less safe method - you use STB and write a macro which takes the generated byte array and passes .data() and .size() so you can pass that in a less generic way to your signature scanner. This could improve times a very little and it'd get rid of that hell-spawn which every other cheat seems to use - for no good reason.

https://github.com/NukoOoOoOoO/winter-linux/blob/1d6a8806d732250ed925d43cb4cfd5f4eb424c7b/src/main.cpp#L17-L32

For stuff like this - in my projects, I developed a "mocking" library, it looks something like this, very rough sketch - written in notepad:
```
struct Mocking {
	Mocking() = delete;
	Mocking(const std::source_location& source_location) : m_source_location(source_location), m_throw(false) {}

	void test() const {
		if (m_throw) {
			// do some custom exception type troll, then catch all of that in the entry-point unless the work you're doing is in another remote thread
			// throw here will be unwinded		
		}
	}

	// now the spicy part, again, this is just a very rough sketch
	void operator,(bool condition) {
		m_throw = !condition;
		test();
	}

	// you might just want to implement some overloads for things like pointers
	// so you can catch nullptrs/custom failure states like unreadable memory page
	// (you can handle that without an operator by making your own pointer class with bool operator() but I'm going offtopic)
	// by "MOCK ptr = (ptr_t*)(...);"
};
```

And in practice it looks something like this:
```
int main() {
	try {
		run_stuff();
	} catch (const MockingException& ex) {
		// this should hold the Mocking stored source_location, and now you handle it
	}
}
```

And when you want to throw:
```
#define DO_MOCK Mocking{std::source_location::current()},

int main() {
	....
	DO_MOCK false;
}
```

https://github.com/NukoOoOoOoO/winter-linux/blob/1d6a8806d732250ed925d43cb4cfd5f4eb424c7b/src/hooks/hooks.cpp#L14

Why not statically store the function pointer with an already well defined type? It'll make things easier - I promise - Helveta abstractizes this quite well (sorry for I keep pointing to Helveta - I just think it could help).

Otherwise, good stuff, I'll look more when I can,

some recommendations:

You might want to look into making your life easier by experiencing with Alternative memory scanning methods - and I have a library exactly for that if you want to try implementing it :)

https://github.com/cristeigabriel/memscan

 I don't plan to officially port it to Linux just yet, I'll soon, but if you plan to do anything with it - given it peaks your interest - that'd be cool. If you want to look for reasons to use it - check the README and tests.

**Comment by NukoOoOoOoO** [comment]
> You can use FindUtf8 directly from localizer. HelvetaCS does this so then you won't have to do UCS->UTF-8. If you need help - I can implement it, although I don't really use Linux for gaming.

Good catch, I didn't about the function, I will use it in the next part of ItemSchema.

> You can just hold instances of all modules as some class' objects and have them inherit stuff like pattern scanner as some sort of traits. You could probably do some concept hackery with this if you want to do it as I said.

> This is unneeded unless you plan on networking strings for signature scanning. You can use STB to compile strings to byte arrays at compile-time. It'd require that you make a template and pass size there to use an array - or - the less safe method - you use STB and write a macro which takes the generated byte array and passes .data() and .size() so you can pass that in a less generic way to your signature scanner. This could improve times a very little and it'd get rid of that hell-spawn which every other cheat seems to use - for no good reason.

> https://github.com/cristeigabriel/memscan

Making pattern scanning better is already in my todo-list but low priority, thanks for pointing out

> For stuff like this - in my projects, I developed a "mocking" library

Well I personally don't see the point of implementing this, but it is good to know some C++20 features <3

> Why not statically store the function pointer with an already well defined type? It'll make things easier - I promise - Helveta abstractizes this quite well (sorry for I keep pointing to Helveta - I just think it could help).

I haven't tried it and I don't know know why, but I will give it a try.

**Comment by cristeigabriel** [comment]
>Well I personally don't see the point of implementing this, but it is good to know some C++20 features <3

To get rid of manual error handling, so it automatically throws in an one liner

```
if (!obj.get()) {
   throw....
}
```

would become

```
MOCK obj.get()
```

**Comment by NukoOoOoOoO** [comment]
Seems good, I will give it a try

**Comment by cristeigabriel** [comment]
If you need help, I can share from my already functional implementation

**Comment by NukoOoOoOoO** [comment]
That would be great if you can help :)

**Comment by cristeigabriel** [comment]
> Oh as for staticalizing original functions, due to the way it is implemented, you cannot make the original function static, otherwise the game will be f×cked

I looked and I see why now. Don't you think that's a bit... not-so-performant? Couldn't you perhaps allocate a page, copy original function there (you can dynamically verify the size of the whole function, perhaps best with a disassembler like Zydis, or the easiest way would be just doing it yourself - just watch out for intersection of bytes, as in, a run-time value could contain what you'd believe is an opcode - and you could accidentally break there), and overwrite at target only once? Then, in your hook, just jump to the stub accordingly. It'd probably make injection time bigger, but it should certainly reduce run-time overhead.

> That would be great if you can help :)
Here's what I do

```
namespace mocking_bird {
	struct exception {
		constexpr exception( const std::source_location& source_location ) : m_source_location( source_location ) { }

		constexpr auto what( ) const {
			return m_source_location;
		}

	private:

		std::source_location m_source_location;
	};

	template< typename T >
	concept Pointer = ( std::is_pointer_v< T > || std::is_same_v< T, nullptr_t > );

	template< typename T >
	concept Operable = requires( T a ) {
		{ a == nullptr } -> std::convertible_to< bool >;
	};

	struct raiser {
		raiser( ) = delete;

		constexpr raiser( const std::source_location& source_location ) : m_source_location( source_location ) { }

		template< typename T >
			requires Pointer< T > && Operable< T >
		constexpr void set( T from ) {
			m_content = ( from != nullptr );
		}

		constexpr void set( bool from ) {
			m_content = from;
		}

		constexpr bool get( ) const {
			return ( m_content == 0 );
		}

		constexpr auto test( ) const {
			return get( ) ? ( std::make_optional( exception( m_source_location ) ) ) : std::nullopt;
		}

		template< typename T >
			requires Pointer< T > && Operable< T >
		constexpr void operator,( T lhs ) {
			operator,( lhs != nullptr );
		}

		constexpr void operator,( bool lhs ) {
			set( lhs );

			const auto ex = test( );
			if ( ex.has_value( ) ) {
				throw ex.value( );
			}
		}

	private:

		bool m_content = false;
		std::source_location m_source_location;
	};
} // namespace mocking_bird

#define MOCK mocking_bird::raiser{ std::source_location::current( ) },
```

**Comment by NukoOoOoOoO** [comment]
Thanks for the code :)
> Don't you think that's a bit... not-so-performant

Well the initial design goal is only make it WORK, never thought about performance and other

> Couldn't you perhaps allocate a page, copy original function there (you can dynamically verify the size of the whole function, perhaps best with a disassembler like Zydis, or the easiest way would be just doing it yourself, and overwrite at target only once? Then, in your hook, just jump to the stub accordingly. It'd probably make injection time bigger, but it should certainly reduce run-time overhead.

Well I have thought about improving it but got delayed by other things, but I will pick it up soon, I am not really satisfied with the current one to be honest

**Comment by cristeigabriel** [comment]
https://github.com/NukoOoOoOoO/winter-linux/blob/478452ea3e6fe3183009590726d1847ef75f22e0/src/menu/menu.cpp#L59

Why would you do this? I mean... it works because address_base_t is a POD, but still

**Comment by NukoOoOoOoO** [comment]
I was blindly pasting that kind of stuff without knowing what I was doing, I am sorry about that:3

**Comment by NukoOoOoOoO** [comment]
![current result](https://i.imgur.com/hrrkIuB.png) So here is the result of the new pattern scanning(by XutaxKamay), ran in release build, and it takes 14~30 ms for each pattern in debug build. I will work on detour next soon, but I need to take a break from it

**Comment by cristeigabriel** [comment]
Beautiful, good stuff :)

**Comment by NukoOoOoOoO** [comment]
Detour is fully implemented now, I also added a small jump method to prevent a situation like the code size of the original function, like CHLClient::LevelInitPostEntity (which is 17 if using long jump method), happens to call other functions at the first 12 bytes

**Comment by cristeigabriel** [comment]
Good stuff
